### PR TITLE
Add BackupSession Controller for Sidecar

### PR DESCRIPTION
**Pull request by suaas21**
**Requirements** : 
- [x]  Merge https://github.com/appscode/stash/issues/673 to master 
- [x]  Merge https://github.com/appscode/stash/issues/675 to master 
- [x] Merge https://github.com/appscode/stash/pull/691
- [x] Rebase against master

**Fixed Issue** : 
- [x]  Add `BackupSession Sidecar Controller` Controller
- [x]  Controller take backup

### CODE REVIEW

File: pkg/backup/backup-session.go [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+
+	// write failure event
+	c.writeBackupFailureEvent(backupSession, err)
+
+	// send prometheus metrics
+	if c.Metrics.Enabled {
+		restoreOutput := &restic.RestoreOutput{
+			SessionDuration: "0s",
+		}
+		return restoreOutput.HandleMetrics(&c.Metrics, err)
```
Comment (hossainemruz) [inline pkg/backup/backup-session.go]: use `backupErr` instead of `err`.

File: pkg/backup/backup-session.go [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+		status.Phase = v1beta1_api.BackupSessionFailed
+		return status
+	}, apis.EnableStatusSubresource)
+
+	// write failure event
+	c.writeBackupFailureEvent(backupSession, err)
+
+	// send prometheus metrics
+	if c.Metrics.Enabled {
+		restoreOutput := &restic.RestoreOutput{
```
Comment (hossainemruz) [inline pkg/backup/backup-session.go]: no need to set `SessionDuration` here.

File: pkg/backup/backup-session.go [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+		eventer.CreateEventWithLog(
+			c.K8sClient,
+			controller.BackupSessionEventComponent,
+			ref,
+			core.EventTypeWarning,
+			eventer.EventReasonBackupSessionFailed,
+			fmt.Sprintf("Failed to restore. Reason: %v", err),
+		)
+	} else {
+		log.Errorf("Failed to restore. Reason: %v", err)
```
Comment (hossainemruz) [inline pkg/backup/backup-session.go]: Use `log.Errorf("Failed to write failure event. Reason: %v", rerr)`

File: pkg/cmds/backup-dir.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+func NewBackupSidecarSessionn() *cobra.Command {
+	con := backup.Controllers{
+		MasterURL:      "",
+		KubeconfigPath: "",
+		Namespace:      meta.Namespace(),
+		MaxNumRequeues: 5,
+		NumThreads:     1,
+		ResyncPeriod:   5 * time.Minute,
+		SetupOpt: restic.SetupOptions{
+			ScratchDir:  "/tmp",
```
Comment (hossainemruz) [inline pkg/cmds/backup-dir.go]: Use constant `ScratchDir`

File: pkg/cmds/backup-dir.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+			EnableCache: false,
+		},
+	}
+
+	cmd := &cobra.Command{
+		Use:               "run-backup",
+		Short:             "Take a backup of workload directory",
```
Comment (hossainemruz) [inline pkg/cmds/backup-dir.go]: `directory` --> `directories`

File: pkg/cmds/backup-dir.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+		DisableAutoGenTag: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			config, err := clientcmd.BuildConfigFromFlags(con.MasterURL, con.KubeconfigPath)
+			if err != nil {
+				log.Fatalf("Could not get Kubernetes config: %s", err)
```
Comment (hossainemruz) [inline pkg/cmds/backup-dir.go]: also return the error

File: pkg/backup/backup-session.go [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+	bsQueue    			 *queue.Worker
+	bsInformer 			 cache.SharedIndexInformer
+	bsLister   			 v1beta1.BackupSessionLister
+
+	Status     			 status.UpdateStatusOptions
+	Recorder   			 record.EventRecorder
+}
+
+func (c *Controllers) RunBackup() error {
+	fmt.Println("RunBackup function")
```
[the remark this answers is not in the document]
Comment (suaas21) [reply pkg/backup/backup-session.go]: fixed

File: pkg/backup/backup-session.go [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	//get BackupConfiguration
+	backupConfiguration, err := c.StashClient.StashV1beta1().BackupConfigurations(c.Namespace).Get(c.Name, metav1.GetOptions{})
+	if err != nil {
+		log.Infoln("fail to get Backup Configuration")
```
[the remark this answers is not in the document]
Comment (suaas21) [reply pkg/backup/backup-session.go]: fixed

File: pkg/cmds/backup-dir.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+			con.StashClient = cs.NewForConfigOrDie(config)
+			con.StashInformerFactory = stashinformers.NewFilteredSharedInformerFactory(
+				con.StashClient,
+				con.ResyncPeriod,
+				con.Namespace,
+				nil,
+			)
+			con.Recorder = eventer.NewEventRecorder(con.K8sClient, backup.BackupEventComponent)
+			if err = con.RunBackup(); err != nil {
+				//set BackupSession status "Failed", write event and prometheus metrics
```
[the remark this answers is not in the document]
Comment (suaas21) [reply pkg/cmds/backup-dir.go]: fixed

File: pkg/backup/backup-session.go [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+		return err
+	}
+
+	// get repository
+	repository, err := c.StashClient.StashV1alpha1().Repositories(backupConfiguration.Namespace).Get(backupConfiguration.Spec.Repository.Name, metav1.GetOptions{})
+	if err != nil {
+		log.Infoln("can't get repository for Backup Configuration")
```
[the remark this answers is not in the document]
Comment (suaas21) [reply pkg/backup/backup-session.go]: fixed

File: pkg/backup/backup-session.go [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+		return err
+	}
+
+	//final configuration of Setupopt and ExtraOpt
+	c.SetupOpt = setupOpt
+	c.ExtraOpt = extraOpt
+
+	// split code from here for leader election
+	if backupConfiguration.Spec.Target == nil {
+		err = fmt.Errorf("backupConfiguration target is nil")
```
[the remark this answers is not in the document]
Comment (suaas21) [reply pkg/backup/backup-session.go]: fixed

File: pkg/backup/backup-session.go (hunk 2 above)
Comment (suaas21) [reply pkg/backup/backup-session.go]: fixed

File: pkg/backup/backup-session.go [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+		return err
+	}
+
+	//get BackupConfiguration for BackupSession
+	backupConfiguration, err := c.StashClient.StashV1beta1().BackupConfigurations(backupSession.Namespace).Get(
+		backupSession.Spec.BackupConfiguration.Name,
+		metav1.GetOptions{},
+	)
+	if err != nil {
+		err := fmt.Errorf("can't get BackupConfiguration for BackupSession %s/%s, reason: %s", backupSession.Namespace, backupSession.Name, err)
```
[the remark this answers is not in the document]
Comment (suaas21) [reply pkg/backup/backup-session.go]: fixed

File: pkg/backup/backup-session.go [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+		if !v {
+			runtime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+			return nil
+		}
+	}
+	c.bsQueue.Run(stopCh)
+	return nil
+}
+
+func (c *Controllers) HandleBackupFailure(err error) error {
```
[the remark this answers is not in the document]
Comment (suaas21) [reply pkg/backup/backup-session.go]: fixed

File: pkg/backup/backup-session.go [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+		return err
+	}
+	// set backupSession phase "Failed"
+	_, err = stash_v1beta1_util.UpdateBackupSessionStatus(c.StashClient.StashV1beta1(), backupSession, func(status *v1beta1_api.BackupSessionStatus) *v1beta1_api.BackupSessionStatus {
+		status.Phase = v1beta1_api.BackupSessionFailed
+		return status
+	}, apis.EnableStatusSubresource)
+
+	// write failure event
+	c.writeBackupFailureEvent(backupSession, err)
```
[the remark this answers is not in the document]
Comment (suaas21) [reply pkg/backup/backup-session.go]: fixed

File: pkg/backup/backup-session.go (hunk 1 above)
Comment (suaas21) [reply pkg/backup/backup-session.go]: fixed

File: pkg/backup/backup-session.go (hunk 3 above)
Comment (suaas21) [reply pkg/backup/backup-session.go]: fixed

File: pkg/cmds/backup-dir.go (hunk 4 above)
Comment (suaas21) [reply pkg/cmds/backup-dir.go]: fixed

File: pkg/cmds/backup-dir.go (hunk 5 above)
Comment (suaas21) [reply pkg/cmds/backup-dir.go]: fixed

File: pkg/cmds/backup-dir.go (hunk 6 above)
Comment (suaas21) [reply pkg/cmds/backup-dir.go]: fixed

File: pkg/backup/backup-session.go [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+	}
+
+	// get repository
+	repository, err := c.StashClient.StashV1alpha1().Repositories(backupConfiguration.Namespace).Get(backupConfiguration.Spec.Repository.Name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	//get host Name
+	host, err := util.GetHostName(backupConfiguration.Spec.Target.Ref)
```
Comment (diptadas) [inline pkg/backup/backup-session.go]: check nil

File: pkg/backup/backup-session.go [hunk 16]
```diff
@@ -0,0 +1,345 @@
...
+	if err != nil {
+		return err
+	}
+
+	//Configure Host, SecretDirectory, EnableCache and ScratchDirectory
+	extraOpt := util.ExtraOptions{
```
Comment (diptadas) [inline pkg/backup/backup-session.go]: assign to c.extraOpt

File: pkg/backup/backup-session.go [hunk 17]
```diff
@@ -0,0 +1,345 @@
...
+		Host:        host,
+		SecretDir:   c.SetupOpt.SecretDir,
+		EnableCache: c.SetupOpt.EnableCache,
+		ScratchDir:  util.ScratchDir,
+	}
+
+	//Configure setupOption
+	setupOpt, err := util.SetupOptionsForRepository(*repository, extraOpt)
```
Comment (diptadas) [inline pkg/backup/backup-session.go]: assign to c.setupOpt

File: pkg/backup/backup-session.go [hunk 18]
```diff
@@ -0,0 +1,345 @@
...
+		backupOpt := util.BackupOptionsForBackupConfig(*backupConfiguration, c.ExtraOpt)
+		backupOutput, err := resticWrapper.RunBackup(&backupOpt)
+		if err != nil {
+			return err
+		}
+
+		//Backup Options configure
+		c.BackupOutput = *backupOutput
+		// If metrics are enabled then generate metrics
+		fmt.Println("=================== Metrics: ",c.Metrics.Enabled,"===================")
```
Comment (diptadas) [inline pkg/backup/backup-session.go]: remove println

File: pkg/backup/backup-session.go [hunk 19]
```diff
@@ -0,0 +1,345 @@
...
+	K8sClient            kubernetes.Interface
+	StashClient          cs.Interface
+	MasterURL            string
+	KubeconfigPath       string
+	StashInformerFactory stashinformers.SharedInformerFactory
+	MaxNumRequeues       int
+	NumThreads           int
+	EnableRBAC           bool // rbac for check job
+	ResyncPeriod         time.Duration
+	//backupConfiguratio
```
Comment (diptadas) [inline pkg/backup/backup-session.go]: comment typo

File: pkg/backup/backup-session.go [hunk 20]
```diff
@@ -0,0 +1,345 @@
...
+	Name      string
```
Comment (diptadas) [inline pkg/backup/backup-session.go]: Name -> BackupConfigurationName

File: pkg/backup/backup-session.go [hunk 21]
```diff
@@ -0,0 +1,345 @@
...
+	Status   status.UpdateStatusOptions
+	Recorder record.EventRecorder
+}
+
+func (c *Controllers) RunBackup() error {
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	//get BackupConfiguration
+	backupConfiguration, err := c.StashClient.StashV1beta1().BackupConfigurations(c.Namespace).Get(c.Name, metav1.GetOptions{})
```
Comment (diptadas) [inline pkg/backup/backup-session.go]: do this after getting backup-session event

File: pkg/backup/backup-session.go (hunk 20 above)
Comment (suaas21) [reply pkg/backup/backup-session.go]: fixed

File: pkg/backup/backup-session.go (hunk 21 above)
Comment (suaas21) [reply pkg/backup/backup-session.go]: fixed

File: pkg/backup/backup-session.go (hunk 18 above)
Comment (suaas21) [reply pkg/backup/backup-session.go]: fixed

File: pkg/backup/backup-session.go (hunk 16 above)
Comment (suaas21) [reply pkg/backup/backup-session.go]: fixed

File: pkg/backup/backup-session.go (hunk 15 above)
Comment (hossainemruz) [reply pkg/backup/backup-session.go]: fixed

File: pkg/backup/backup-session.go (hunk 19 above)
Comment (hossainemruz) [reply pkg/backup/backup-session.go]: fixed

File: pkg/backup/backup-session.go (hunk 17 above)
Comment (hossainemruz) [reply pkg/backup/backup-session.go]: fixed

File: pkg/backup/backup-session.go [hunk 22]
```diff
@@ -0,0 +1,361 @@
...
+
+func (c *Controllers) electBackupLeader(backupConfiguration *v1beta1_api.BackupConfiguration, stopCh <-chan struct{}) error {
+	rlc := resourcelock.ResourceLockConfig{
+		Identity:      os.Getenv(util.KeyPodName),
+		EventRecorder: eventer.NewEventRecorder(c.K8sClient, BackupEventComponent),
+	}
+	resLock, err := resourcelock.New(
+		resourcelock.ConfigMapsResourceLock,
+		backupConfiguration.Namespace,
+		util.GetConfigmapLockName(api.LocalTypedReference{
```
[the remark this answers is not in the document]
Comment (hossainemruz) [reply pkg/backup/backup-session.go]: fixed

File: pkg/controller/backup_session.go [hunk 23]
```diff
@@ -80,12 +81,19 @@ func (c *StashController) runBackupSessionProcessor(key string) error {
 		backupSession := obj.(*api_v1beta1.BackupSession)
```
Comment (diptadas) [inline pkg/controller/backup_session.go]: don't need else
`return c.setBackupSessionRunning(backupSession)` at the end

File: pkg/controller/backup_session.go [hunk 24]
```diff
@@ -257,3 +279,60 @@ func (c *StashController) setBackupSessionRunning(backupSession *api_v1beta1.Bac
...
+	_, err := stash_util.UpdateBackupSessionStatus(c.stashClient.StashV1beta1(), backupSession, func(in *api_v1beta1.BackupSessionStatus) *api_v1beta1.BackupSessionStatus {
+		in.Phase = api_v1beta1.BackupSessionSucceeded
+		in.SessionDuration = sessionDuration.String()
+		return in
+	}, apis.EnableStatusSubresource)
+	if err != nil {
+		return err
+	}
+
+	// write job creation success event
```
Comment (diptadas) [inline pkg/controller/backup_session.go]: fix comment

File: pkg/util/options.go [hunk 25]
```diff
@@ -36,20 +36,22 @@ func RestoreOptionForRestoreSession(restoreSession api.RestoreSession, extraOpt
...
 	for _, rule := range rules {
 		// if host is specified in rule then use it. otherwise use workload itself as host
-		targetHost := hostname
-		if rule.Host != "" {
-			targetHost = rule.Host
+		sourceHost := hostname
+		if rule.SourceHost != "" {
+			sourceHost = rule.SourceHost
 		}
 		if len(rule.Subjects) == 0 {
```
Comment (diptadas) [inline pkg/util/options.go]: combine `len(rule.Subjects) == 0 || go_str.Contains(rule.Subjects, hostname)`

File: apis/util.go [hunk 26]
```diff
@@ -24,4 +24,5 @@ const (
 	KindDaemonSet             = "DaemonSet"
 	KindPersistentVolumeClaim = "PersistentVolumeClaim"
 	KindAppBinding            = "AppBinding"
+	ScratchDir                = "/tmp"
```
[the remark this answers is not in the document]
Comment (hossainemruz) [reply apis/util.go]: used

File: pkg/controller/backup_session.go (hunk 24 above)
Comment (hossainemruz) [reply pkg/controller/backup_session.go]: fixed

File: pkg/controller/backup_session.go (hunk 23 above)
Comment (hossainemruz) [reply pkg/controller/backup_session.go]: fixed

File: pkg/util/options.go (hunk 25 above)
Comment (hossainemruz) [reply pkg/util/options.go]: fixed

**Pull request merged**